optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +127,134 @@ private void initBenchmarkDirectoryAndSubdirs() {
+        // Wait for the warm up benchmarks to complete
+        int tasksCount = futureMap.size();
+        for (int i = 0; i < tasksCount; i++) {
+            Future<SingleBenchmarkRunner> future;
             try {
-                Solver solver = solverBenchmarkResult.getSolverConfig().buildSolver();
-                solver.solve(problemBenchmarkResult.readPlanningProblem());
-            } catch (RuntimeException e) {
-                singleBenchmarkRunnerExceptionLogger.warn("The warmUp of singleBenchmark ("
-                        + singleBenchmarkResult.getName() + ") failed.", e);
-                // TODO update firstFailureSingleBenchmarkRunner (when warmups happen in a Runner too in parallel)
+                future = warmUpExecutorCompletionService.take();
+            } catch (InterruptedException e) {
+                singleBenchmarkRunnerExceptionLogger.error("Waiting for a warm up singleBenchmarkRunner was interrupted.", e);
+                continue;
             }
 
-            solverBenchmarkResult.getSolverConfig().setTerminationConfig(originalTerminationConfig);
-            long timeSpent = System.currentTimeMillis() - startingTimeMillis;
-            timeLeft = warmUpTimeMillisSpentLimit - timeSpent;
-            if (timeLeft <= 0L) {
-                return timeLeft;
+            Throwable failureThrowable = null;
+            SingleBenchmarkRunner singleBenchmarkRunner = futureMap.get(future);
+            try {
+                // Explicitly returning it in the Callable guarantees memory visibility
+                singleBenchmarkRunner = future.get();
+            } catch (InterruptedException e) {
+                singleBenchmarkRunnerExceptionLogger.error("The warm up singleBenchmarkRunner ({}) was interrupted.",
+                        singleBenchmarkRunner, e);
+                failureThrowable = e;
+            } catch (ExecutionException e) {
+                Throwable cause = e.getCause();
+                singleBenchmarkRunnerExceptionLogger.warn("The warm up singleBenchmarkRunner ({}) failed.",
+                        singleBenchmarkRunner, cause);
+                failureThrowable = cause;
+            }
+            if (failureThrowable != null) {
+                singleBenchmarkRunner.setFailureThrowable(failureThrowable);
+                if (firstFailureSingleBenchmarkRunner == null) {
+                    firstFailureSingleBenchmarkRunner = singleBenchmarkRunner;
+                }
+            }
+
+            SolverBenchmarkResult solverBenchmarkResult = singleBenchmarkRunner.getSingleBenchmarkResult().getSolverBenchmarkResult();
+            long timeLeftInCycle = timePhaseEnd - System.currentTimeMillis();
+            if (timeLeftInCycle > 0L) {
+                warmUpPopulate(futureMap, singleBenchmarkResultIndexMap, new SolverBenchmarkResult[]{solverBenchmarkResult}, timeLeftInCycle);
+                tasksCount++;
```

Comment:
Isn't it possible that the warmUpPopulate adds more than just 1 Future into the futureMap? so tasksCounts++ might not be accurate.

I think the code would be more readable if the for(i to taskCounts) loop would be replaced by a while (futureMap isn't empty, pop the first one)
Note: pop() is a method on Queue. It can be mimiced on Map with values().iterator().next(), but a Queue is probably more natural here?